Add PiP(Picture in Picture) mode to the player

suhailkakar opened the pull request:
## Description

Added picture in picture mode to the Player (demo below)

#51 

## Additional Information

- [X] I read the [contributing docs](/livepeer/livepeer.js/blob/main/.github/CONTRIBUTING.md) (if this is your first contribution)

## Demo

https://user-images.githubusercontent.com/56798748/196104524-bea638a9-d148-4e0f-bc37-c5efb130e7b5.mov

suhailkakar committed:
feat: added PiP to the player

adamsoffer reviewed (changes requested):
This looks great! Let's resolve those two eslint failures then good to merge after Chase's review.

0xcadams commented:
This is a great start! Some things I'd like to see on this:

- [x] the general architecture (needs documentation) of the video controls is that the store is updated directly and maintains state, and then we add side effects to that store which interacts with the media element based on the diff between states. There should be a good example with Fullscreen which shows this - the button triggers the modification of the state, and then there is a listener which calls functions on the element. Let's change this to handle PiP similarly. I will document this better in the future when I'm back.
- [x] We will need to test on different browsers and I would generally follow how mux implemented their PiP here (but not directly, since they handle some things we might not care about) https://github.com/muxinc/media-chrome/blob/main/src/js/media-controller.js#L168 I think this PiP request would cause a crash with older browsers as-is, so we'll need to update this
- [x] Let's add a changeset to this PR which describes the updates made (see historical PRs for examples) to create a new changeset file, run `pnpm changeset`
- [x] let's export this component from the main index.ts, so bubble up that export until it's exported from the main index.ts. so someone could use this component in their own custom controls if they wanted to
- [ ] This is more of a question, but should this be included by default with the player? Or should it be, by default, not shown, but optional with some kind of flag (showPipButton)? I'm leaning towards the latter but would love feedback on it!

Thanks for this PR!

0xcadams reviewed (changes requested):
See above for details

0xcadams commented:
Also, to help fix the types, we'll need to update this file with PiP functions. There's some documentation in that global types file which explains what those custom types are, but basically it's typing how historical browsers have defined a "Fullscreen request" in the past. We may not need to add the multiple definitions of "Fullscreen request", depending on what the history is of PiP (I don't know how safari or Firefox have evolved over the years to handle this)
https://github.com/livepeer/livepeer.js/blob/main/global.d.ts

suhailkakar committed:
refactor: updated the pip architecture

suhailkakar commented:
- Updated the architecture of the picture in picture mode (now it is similar to full screen - store is updated directly and maintains state)

- Exported the component and added changset

- The `requestPictureInPicture` API is not supported by Firefox, therefore to avoid crashes, I have added a condition to not render the picture in picture button if the browser doesn't support it. 

<img width="779" alt="image" src="https://user-images.githubusercontent.com/56798748/196410874-01a67cef-20b4-4ec9-83bd-ecb2824f22b0.png">

<img width="1656" alt="image" src="https://user-images.githubusercontent.com/56798748/196412658-55d0e61f-3d1f-4056-9201-07080ed6b453.png">



> This is more of a question, but should this be included by default with the player? Or should it be, by default, not shown, but optional with some kind of flag (showPipButton)? I'm leaning towards the latter but would love feedback on it!

I think it is better to show the button by default but if the users want they can add `disabledPiP={false}` to remove it.

suhailkakar reviewed (commented):
Fixed the lint issue

0xcadams commented on .changeset/poor-trains-divide.md:
```diff
@@ -0,0 +1,5 @@
+---
+'@livepeer/react': major
```
Let's change this to be `minor`

suhailkakar replied:
Sure, thanks : [#1e81aef](https://github.com/livepeer/livepeer.js/pull/98/commits/1e81aefbbe99698ced5e03a186e32c8c208e7fce)

0xcadams commented on global.d.ts:
```diff
@@ -25,6 +25,10 @@ declare global {
+    /**
+     * Enters picture-in-picture mode
+     */
+    requestPictureInPicture?(): void;
```
We'll need to add a few more methods here and handle the outdated browsers. It's annoying, but we promise 94% of browsers and we can easily add checks for the existence of WebKit.

https://caniuse.com/?search=picture-in-picture

We can merge this and I will create a follow-up task to make sure we're handling those legacy browsers.

0xcadams commented on packages/core/src/media/controls/controls.ts:
```diff
@@ -165,6 +181,9 @@ export const createControllerStore = <TElement extends HTMLMediaElement>(
         fullscreen: false,
+        pictureInPicture: false,
+
+        mail: {},
```
Is this used anywhere?

suhailkakar replied:
Sorry, it was just a typo :)

0xcadams replied:
Sounds good!

suhailkakar replied:
Fixed typo: [#34ceffa](https://github.com/livepeer/livepeer.js/pull/98/commits/34ceffaf2b9cbd83a14094987bd52c2ce373ba87)

0xcadams commented on packages/core/src/media/controls/pictureinpicture.ts:
```diff
@@ -0,0 +1,116 @@
+  const { methods } = getPictureInPictureMethods();
+
+  if (methods) {
+    document.addEventListener(methods.exitPictureInPicture, callback, false);
```
I don't think this is an event
https://developer.mozilla.org/en-US/docs/Web/API/Picture-in-Picture_API#events

suhailkakar replied:
Yep, `exitPictureInPictureMode` is a separate API: https://developer.mozilla.org/en-US/docs/Web/API/Document/exitPictureInPicture

0xcadams replied:
Hmm but we're adding an event listener here and that event name doesn't exist

suhailkakar replied:
I see, then I believe it is better to use `leavepictureinpicture`

https://developer.mozilla.org/en-US/docs/Web/API/HTMLVideoElement/leavepictureinpicture_event

0xcadams commented on packages/core/src/media/controls/controls.ts:
```diff
@@ -165,6 +181,9 @@ export const createControllerStore = <TElement extends HTMLMediaElement>(
         hasPlayed: false,
         playing: !element?.paused,
         fullscreen: false,
+        pictureInPicture: false,
```
Let's make sure this value is being updated properly on enter/exit of PiP. Users can consume this value to conditionally render UI when PiP is active, so we need to make sure it toggles correctly. I can't test locally currently so just leaving this here to make sure it's tested

0xcadams commented on packages/core/src/media/controls/pictureinpicture.ts:
```diff
@@ -0,0 +1,116 @@
+const getPictureInPictureMethods = (element?: HTMLMediaElement | null) => {
+  if (isClient()) {
+    for (const methods of methodsList) {
+      const exitPictureInPicturenMethod = methods.exitPictureInPicture;
```
Typo here

suhailkakar replied:
Fixed typo: [#34ceffa](https://github.com/livepeer/livepeer.js/pull/98/commits/34ceffaf2b9cbd83a14094987bd52c2ce373ba87)

0xcadams commented on packages/react/src/components/media/controls/PictureInPictureButton.tsx:
```diff
@@ -0,0 +1,92 @@
+  const _children = icon ? icon : <DefaultPictureInPictureIcon />;
+
+  const isPiPSupported = isPictureInPictureSupported(
+    document.querySelector('video'),
```
Let's use the _element value from the `useMediaController` hook above instead of this document query.

suhailkakar replied:
Sure: [#2a2bd5a](https://github.com/livepeer/livepeer.js/pull/98/commits/2a2bd5ad57bb2c18e7158cdbe230272ff68d4c09)

0xcadams reviewed (changes requested):
This looks awesome!! A few more requests please. And I'm not sure I see the reasoning for including PiP by default - it seems like a niche feature that some apps will want, but most won't. I personally don't see it on most video apps I use. It seems to me that it should default to not shown with a documented flag to enable it. Let me know your thoughts!

0xcadams commented:
Also, can we document this somewhere in the Player docs?

0xcadams committed:
Merge branch 'main' into suhailkakar/PiP-mode

suhailkakar committed:
chore: updated changset to minor

suhailkakar committed:
Merge branch 'suhailkakar/PiP-mode' of https://github.com/suhailkakar/livepeer.js into suhailkakar/PiP-mode

suhailkakar committed:
refactor: used _element instead of query selector on picture in picture button

suhailkakar committed:
feat: added PiP props to player

suhailkakar committed:
fixed typos

suhailkakar committed:
fix: updated listeners for PiP

suhailkakar committed:
refactor: replaced relative imports with absolute imports

suhailkakar commented:
> Let's make sure this value is being updated properly on enter/exit of PiP. Users can consume this value to conditionally render UI when PiP is active, so we need to make sure it toggles correctly. I can't test locally currently so just leaving this here to make sure it's tested

@0xcadams I just checked but non of the variables such as full screen, and volume are updating their values from stores.

0xcadams committed:
fix: added old webkit support and docs

0xcadams committed:
docs: fixed dev example

0xcadams commented:
> > Let's make sure this value is being updated properly on enter/exit of PiP. Users can consume this value to conditionally render UI when PiP is active, so we need to make sure it toggles correctly. I can't test locally currently so just leaving this here to make sure it's tested
> 
> @0xcadams I just checked but non of the variables such as full screen, and volume are updating their values from stores.

True!! Nice catch, thanks for that. Just pushed a fix for that and added old webkit and documentation. Once the CI passes we shall 🚢 🚢 🚢

0xcadams commented:
Merging and seeing if tests pass in `main`.
